docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+     */
+    public long multipartDownloadthreshold();
+
+    /**
+     * The maximum number of parts objects are to be downloaded in.
+     */
+    public int maximumDownloadPartCount();
+
+    /**
+     * The minimum size for each part.
+     */
+    public long minimumDownloadPartSize();
+}
+
+/**
+ * Configuration object for multipart uploads.
+ */
+public final class MultipartUploadConfiguration {
+    /**
+     * Whether multipart uploads should be enabled.
+     */
+    public boolean enableMultipartUpload();
+
+    /**
+     * The minimum size for an object to be uploaded in multipe parts.
+     */
+    public long multipartUploadThreshold();
+
+    /**
+     * The minimum size for each uploaded part.
+     */
+    public long minimumUploadPartSize();
+}
+
+/**
+ * Override configuration for a single transfer.
+ */
+public interface TransferOverrideConfiguration {
+    /**
+     * The maximum rate for this transfer in bytes per second.
+     */
+    Optional<Long> maxTransferBytesPerSecond();
+
+    /**
+     * Override configuration for multipart downloads.
+     */
+    Optional<MultipartDownloadConfiguration> multipartDownloadConfiguration();
+
+    /**
+     * Override configuration for multipart uploads.
+     */
+    Optional<MultipartUploadConfiguration> multipartUploadConfiguration();
+}
+
+
+/**
+ * A factory capable of creating the streams for individual parts of a given
+ * object to be uploaded to S3.
+ * <p>
+ * There is no ordering guaranatee for when {@link
+ * #streamForPart(PartUploadContext)} is called.
+ */
+public interface ObjectPartStreamCreator {
+    /**
+     * Return the stream for the object part described by given {@link
+     * PartUploadContext}.
+     *
+     * @param context The context describing the part to be uploaded.
+     * @return The part stream.
+     */
+    Publisher<ByteBuffer> streamForPart(PartUploadContext context);
+
+    /**
+     * Return the stream for a entire object to be uploaded as a single part.
+     */
+    Publisher<ByteBuffer> streamForObject(SinglePartUploadContext context);
+
+    /**
+     * Create a factory that creates streams for individual parts of the given
+     * file.
+     *
+     * @param file The file whose parts the factory will create streams for.
+     * @return The stream factory.
+     */
+    static ObjectPartStreamCreator forFile(Path file) {
+        return ...;
+    }
+}
+
+/**
+ * A factory capable of creating the {@link AsyncResponseTransformer} to handle
+ * each downloaded object part.
+ * <p>
+ * There is no ordering guarantee for when {@link
+ * #transformerForPart(PartDownloadContext)} invocations. It is invoked when the
+ * response from S3 is received for the given part.
+ */
+public interface ObjectPartTransformerCreator {
+    /**
+     * Return a transformer for downloading a single part of an object.
+     */
+    AsyncResponseTransformer<GetObjectResponse, ?> transformerForPart(MultipartDownloadContext context);
```

Comment:
Does the creator need to be parameterized with the output type?

Reply:
Not in the way it's currently used since the `Transfer` object (or subclasses) doesn't expose the output from the transformer.

Reply:
Then are we using the wrong abstraction? If the output of the transformation doesn't matter, does it need to be a transformation, or is it just a `Consumer`?

Reply:
It's possible, but I was hesitant to simplify it since the `GetObjectResponse` given to the `AsyncResponseTransformer` may be important to the implementation, as well as the retry behavior you get by completing the future returned from `prepare()` appropriately.

Reply:
Cool, we can always work it out later.